### PR TITLE
Implement UDP transmission

### PR DESCRIPTION
**Pull request by JacksonCoder**
This PR adds a UDP client to the application, which sends a JSON-encoded description of the robot position to the roboRIO.

Closes https://github.com/SouthEugeneRoboticsTeam/dunk/issues/1

### CODE REVIEW

File: UDPClient.cpp [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+void UDPClient::send(string msg) {
+    if (!connected) return;
+    auto msg_str = msg.c_str();
+    if (sendto(socket_i, msg_str, strlen(msg_str)+1, 0, // +1 to include terminator
+               (sockaddr*)&roborio_addr, sizeof(roborio_addr)) < 0){
+        perror("Could not send UDP message");
+        close(socket_i);
+        connected = false;
+    }
+}
```
Comment (andrewda) [inline UDPClient.cpp]: EOF

File: UDPClient.h [hunk 2]
```diff
@@ -0,0 +1,32 @@
+//
+// Created by jacksoncoder on 9/4/18.
+//
+
+#ifndef DUNK_SERVER_H
+#define DUNK_SERVER_H
+
+
+
```
Comment (andrewda) [inline UDPClient.h]: Too many newlines (just use 1)

File: UDPClient.h [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+    sockaddr_in roborio_addr;
+    int socket_i;
+    bool connected = false;
+
+    public:
+        UDPClient(string host, uint16_t port);
+        void send(string msg);
+};
+
+
```
Comment (andrewda) [inline UDPClient.h]: 1 newline

File: json.cpp [hunk 4]
```diff
@@ -0,0 +1,10 @@
+#include "json.h"
+
+string build_json_msg(double x, double y, double angle) {
+    string tmp = "{";
+    tmp += string( "\"x\":") + to_string(x);
```
Comment (andrewda) [inline json.cpp]: don't wrap x, y, angle in quotations in the JSON object, they are numbers and we want kotlin to interpret them as such

Comment (andrewda) [reply json.cpp]: whoops, this is fine as-is

File: json.cpp [hunk 5]
```diff
@@ -0,0 +1,10 @@
+#include "json.h"
+
+string build_json_msg(double x, double y, double angle) {
+    string tmp = "{";
```
Comment (andrewda) [inline json.cpp]: maybe a more descriptive variable name here, cause why not. `json` would be fine, or `jsonMessage` or what have you

**Comment by andrewda** [comment]
Oh @JacksonCoder one more thing - could you send the timestamp with the JSON?

**Comment by andrewda** [comment]
Down to the milliseconds

**Comment by JacksonCoder** [comment]
Sure!

**Comment by andrewda** [comment]
ping @JacksonCoder

**Comment by JacksonCoder** [comment]
Right, forgot about this. Will fix in 5 minutes.

**Comment by JacksonCoder** [comment]
Ahhh, stupid merge conflicts.

**Comment by JacksonCoder** [comment]
Alright, there we go.

**Comment by JacksonCoder** [comment]
In this scenario, `5 == 25`.

File: json.cpp [hunk 6]
```diff
@@ -0,0 +1,18 @@
+#include "json.h"
+
+string json_message(double x, double y, double angle) {
+  // Get time
+  struct timeval tp;
+  gettimeofday(&tp, NULL);
+  long int epoch_s = tp.tv_sec;
+  long int epoch_ms = tp.tv_usec * 1000;  
+  string json = "{";
+  json += string(" \"epoch_s\": ") + to_string(epoch_s);
```
Comment (andrewda) [inline json.cpp]: We don't need seconds if we're doing ms lol. I think Kotlin is capable of dividing by 1000

Comment (JacksonCoder) [reply json.cpp]: These would be the sub-milliseconds, not the total milliseconds.

Comment (andrewda) [reply json.cpp]: Oh... okay then

**Pull request merged**